Bluetooth: Host: Fix L2CAP EINPROGRESS refs

alwa-nordic opened the pull request:
Fix discrepancy in reference management between calls to `bt_l2cap_chan_ops.recv` when the application returns `-EINPROGRESS`.

There are two call sites, `l2cap_chan_le_recv_sdu` and `l2cap_chan_le_recv`, that were inconsistent.

`l2cap_chan_le_recv_sdu` moves the reference, and this patch updates `l2cap_chan_le_recv_sdu` to do the same.

This behavior is also now documented.

This bug has existed since the introduction of this feature in 3151d265726fb528d31d8e1ee8b35004e901cfc4.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Fix `bt_l2cap_chan_ops.recv` `-EINPROGRESS`

Fix discrepancy in reference management between calls to
`bt_l2cap_chan_ops.recv` when the application returns `-EINPROGRESS`.

There are two call sites, `l2cap_chan_le_recv_sdu` and
`l2cap_chan_le_recv`, that were inconsistent.

`l2cap_chan_le_recv_sdu` moves the reference, and this patch updates
`l2cap_chan_le_recv` to do the same.

This behavior is also now documented.

This bug has existed since the introduction of this feature in
3151d265726fb528d31d8e1ee8b35004e901cfc4.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Test L2CAP -EINPROGRESS feature

The test implementation is based on a copy of the HFC multilink test.
The test verifies that the stack respects the reference counting of SDU
buffers when the L2CAP -EINPROGRESS feature is used.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2611,7 +2613,14 @@ static void l2cap_chan_le_recv(struct bt_l2cap_le_chan *chan,
+		net_buf_unref(owned_ref);
+		owned_ref = NULL;
+	}
+
+
```
minor: drop the other empty line

jori-nordic reviewed (commented):
Please also clean-up and submit the test that found the bug.

jhedberg commented:
GitHub doesn't allow me to do inline review of commit messages, so I'll do it here:

> `l2cap_chan_le_recv_sdu` moves the reference, and this patch updates
> `l2cap_chan_le_recv_sdu` to do the same.

The second function reference there should be `l2cap_chan_le_recv`, shouldn't it?

alwa-nordic commented:
Added test and fixed the commented-on issues.

jori-nordic commented on tests/bsim/bluetooth/host/l2cap/einprogress/data.h:
```diff
@@ -0,0 +1,20 @@
+ */
+
+#ifndef ZEPHYR_TESTS_BSIM_BLUETOOTH_HOST_MISC_HFC_MULTILINK_SRC_DATA_H_
+#define ZEPHYR_TESTS_BSIM_BLUETOOTH_HOST_MISC_HFC_MULTILINK_SRC_DATA_H_
```
s/hfc_multilink/something-else

jori-nordic commented on tests/bsim/bluetooth/host/l2cap/einprogress/dut/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(hfc_multilink)
```
s/hfc_multilink/something-else

jori-nordic commented on tests/bsim/bluetooth/host/l2cap/einprogress/tester/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(hfc_multilink_tester)
```
hfc_multilink

jori-nordic commented on tests/bsim/bluetooth/host/misc/hfc_multilink/dut/src/dut.c:
```diff
@@ -26,15 +30,43 @@ LOG_MODULE_REGISTER(dut, CONFIG_APP_LOG_LEVEL);
 
-struct tester {
+struct tester_t {
 	size_t sdu_count;
```
please revert changes to `hfc_multilink`

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2611,7 +2613,13 @@ static void l2cap_chan_le_recv(struct bt_l2cap_le_chan *chan,
 
-	err = chan->chan.ops->recv(&chan->chan, buf);
+	owned_ref = net_buf_ref(buf);
+	err = chan->chan.ops->recv(&chan->chan, owned_ref);
```
What if `chan->chan.ops->recv == NULL`?

alwa-nordic replied:
Out of scope of this PR?

Thalley replied:
Trivial fix though - Just apply the same solution as we have for `alloc_buf` on line 2589.

alwa-nordic replied:
But unlike `alloc_buf`, `recv` is mandatory. The only way for this to fail is if the application messes with the `ops` struct when it shouldn't. That's a serious error that might as well trigger a segfault. There is no bug here.

We could be nice and recover. But there are many ways. Just substitute in a no-op `recv`, like you say? But then the application does not know it broke something. Disconnect the channel?

Defensive code is nice, but I would prefer this is discussed in a separate issue. Then @jori-nordic and @jhedberg can give their input. Will you make an issue @Thalley?

Thalley replied:
> But unlike alloc_buf, recv is mandatory.

Where is that documented? Implied by the lack of `if` in the description?

> The only way for this to fail is if the application messes with the ops struct when it shouldn't. That's a serious error that might as well trigger a segfault. There is no bug here.

Why shouldn't the application modify this? It's application owned, isn't it? 

> Just substitute in a no-op recv, like you say? But then the application does not know it broke something. 

Not sure if I would consider that broken. It's (IMO) not much different than not providing the `recv` callback for scanning. 

I'm not understanding the reluctance of checking if a pointer is NULL before dereferencing it

Thalley replied:
> here my good sir

I forgot about that. I was too busy looking in the documentation which doesn't describe this requirement (it probably should, but that's for another PR). 

> An __ASSERT_NO_MSG() maybe, but I wouldn't add error handling.

Since this would be due to an application, I don't think an assert is the right course here. 
`LOG_ERR` maybe? Failing on segmentation fault just seem like a subpar approach

jori-nordic replied:
The problem is that error logs are not actionable by said app. So say it malloc'd the ops struct and accidentally frees it after zeroing it out. It won't know that the channel is still alive and receiving data.

I agree that this situation really sucks because asserts are disabled in prod. So the app won't know in either case. Maybe a `BT_ASSERT()`?

jori-nordic replied:
@Thalley https://github.com/zephyrproject-rtos/zephyr/pull/77115

Thalley replied:
> The problem is that error logs are not actionable by said app.

Well neither are asserts, right? Both can give information to the developer, but the app itself cannot do anything, so in that regard they are equally good, no? 

Of course NULL checks won't fix all memory issues either as `recv = 0x1234` is non-NULL but would still break stuff. 

At this point I am OK with any solution, but not checking an application provided pointer before dereferencing is not great :)

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/dut/src/dut.c:
```diff
@@ -0,0 +1,228 @@
+
+LOG_MODULE_REGISTER(dut, CONFIG_APP_LOG_LEVEL);
+
+struct tester_t {
```
What's the `_t` in `tester_t`?

alwa-nordic replied:
This is copy-pasted code from a different test. There it was just `tester`. But the CI now dislikes overlap in the struct type namespace and variable namespace. I just added `_t` for "type". Any other ideas?

Thalley replied:
> I just added _t for "type"

Our coding guidelines prohibit that :) 

Could just be something like `tester_data` or `tester_inst` or something

jhedberg replied:
> > I just added _t for "type"
> 
> Our coding guidelines prohibit that :)

Just to follow up on this. We don have a convention of `_t` but that's for typedefs, like `bt_addr_le_t`. With a struct name you either way have to do `struct foo` whenever you reference the type, so it's redundant to put any additional stuff in the name itself.

jori-nordic replied:
Where is that rule documented?
https://kernel.org/doc/html/latest/process/coding-style.html#typedefs only talks about typedefs vs `struct foo`, not the `struct foo_t` being illegal.

jhedberg replied:
> Where is that rule documented?

Not sure if that was a question to me or @Thalley, but I don't think I said "rule" anywhere. I was only talking about the existing convention (as well as explaining the rationale behind it). That said, do we really need to have an explicitly documented "Be consistent with existing code" rule? I'd have hoped that it goes without saying.

Thalley replied:
> Where is that rule documented? https://kernel.org/doc/html/latest/process/coding-style.html#typedefs only talks about typedefs vs `struct foo`, not the `struct foo_t` being illegal.

Indeed. It only applies to `typedefs` as has been pointed out. I made a mistake in assuming that the same idea would also apply for structs. 

> That said, do we really need to have an explicitly documented "Be consistent with existing code" rule? I'd have hoped that it goes without saying.

We actually used to have that somewhere IIRC back in 2018-2019-ish, but I'm failing to find it now.

jori-nordic replied:
> Be consistent with existing code

Since this existing code style differs between subsystems and sometimes even between individual units, I would like to have a document stating what the policy is.

I'm being obtuse on purpose, because the process people are burying their head in the sand with #75772. If we had guidelines, this whole thread wouldn't exist.

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/dut/src/dut.c:
```diff
@@ -0,0 +1,228 @@
+{
+	int buf_id = net_buf_id(destroyed_buf);
+
+	__ASSERT_NO_MSG(0 <= buf_id && buf_id < ARRAY_SIZE(acl_pool_refs_held));
```
Why `__ASSERT_NO_MSG` instead of `TEST_ASSERT`?

alwa-nordic replied:
I'm trying to be helpful in locating the problem for anyone debugging.

The assert is an assertion in the local code, i.e. the problem is in test itself.

OTOH `TEST_ASSERT` is checking some property about the DUT. If it fails, the DUT is not following the specification encoded by the test.

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/tester/prj.conf:
```diff
@@ -0,0 +1,19 @@
+CONFIG_BT_BUF_CMD_TX_SIZE=255
+CONFIG_BT_BUF_EVT_DISCARDABLE_SIZE=255
+
+# Work around bug in throughput calculation
+# FIXME: check if this still applies
+CONFIG_BT_CTLR_FORCE_MD_AUTO=n
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
```
What bug is this?

alwa-nordic replied:
@jori-nordic

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/tester/src/tester.c:
```diff
@@ -0,0 +1,675 @@
+	net_buf_simple_save(&buf->b, &state);
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+
+	if (hdr->evt == BT_HCI_EVT_CMD_COMPLETE) {
+		struct bt_hci_evt_cmd_complete *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = 0;
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else if (hdr->evt == BT_HCI_EVT_CMD_STATUS) {
+		struct bt_hci_evt_cmd_status *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = buf->data[0];
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else {
+		__ASSERT_NO_MSG(0);
+	}
+
+	LOG_DBG("opcode 0x%04x status %x", opcode, status);
+
+	__ASSERT(status == 0x00, "cmd status: %x", status);
+
+	__ASSERT(active_opcode == opcode, "unexpected opcode %x != %x", active_opcode, opcode);
+
+	if (active_opcode) {
+		active_opcode = 0xFFFF;
+		cmd_rsp = net_buf_ref(buf);
+		net_buf_simple_restore(&buf->b, &state);
+	}
+
+	if (ncmd) {
+		k_sem_give(&cmd_sem);
+	}
+}
```
Why are we handling this and more in this file? We can't test everything just via the L2CAP API?

jori-nordic replied:
![image](https://github.com/user-attachments/assets/9a73e29c-21aa-4c61-bd66-67d64c090135)

Thalley replied:
![image](https://github.com/user-attachments/assets/3d26eb86-eb68-4978-9016-0a1e218e89c4)

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/tester/src/tester.c:
```diff
@@ -0,0 +1,675 @@
+	}
+
+	k_sem_take(&cmd_sem, K_FOREVER);
+	__ASSERT_NO_MSG(active_opcode == 0xFFFF);
```
Suggest to define `0xFFFF` to not use magic numbers

alwa-nordic commented:
I'll delete the tester code, and rewrite the test. Basing it on a different test was a mistake.

alwa-nordic committed:
Bluetooth: Host: Upgrade log severity for L2CAP user error

For ease of development, we should log the event as an error.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Add BT_TESTING trace event for ACL pool destroy

This is needed for a test to catch a double-free.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: testlib: Add `BT_TESTLIB_ADDR_LE_RANDOM_C0_00_00_00_00_`

This is shorthand for random static addresses. It's similar to
`bt_addr_le_from_str`, but is a macro that results in an object literal,
making it more versatile and less verbose.

This macro only gives access to the first 255 random static addresses,
but this ought to be enough addresses for testing.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -349,6 +349,9 @@ struct bt_l2cap_chan_ops {
 	 *  @kconfig{CONFIG_BT_L2CAP_SEG_RECV} is enabled and seg_recv is
 	 *  supplied.
 	 *
+	 *  If the application returns -EINPROGRESS, the reference
+	 *  @p buf was moved.
```
What does `moved` mean in this context?

jori-nordic replied:
@alwa-nordic that means the note is not clear enough. 
@Thalley it means the reference was moved into the application. The app now owns `buf`.

Thalley commented on tests/bsim/bluetooth/host/l2cap/einprogress/compile.sh:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+set -eu
+: "${ZEPHYR_BASE:?ZEPHYR_BASE must be defined}"
+
+INCR_BUILD=1
```
When should or shouldn't we use `INCR_BUILD` in BSIM tests?

Thalley replied:
Good to know :) Thanks

Thalley reviewed (dismissed):
No blockers from me

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -262,13 +263,23 @@ void bt_send_one_host_num_completed_packets(uint16_t handle)
 }
 
+#if defined(CONFIG_BT_TESTING)
+__weak void bt_testing_trace_event_acl_pool_destroy(struct net_buf *buf)
```
Seems like we might need a `include/zephyr/bluetooth/testing.h` which would contain the various prototypes of testing-related functions and also properly document them. I noticed we have some existing cases that could use it, like here:
https://github.com/zephyrproject-rtos/zephyr/blob/1471c2f81430fd0a381c37d263a4d0ff966e0cb4/tests/bsim/bluetooth/host/misc/disconnect/dut/src/main.c#L32-L33

Thalley replied:
Do we want to expose test functions in a public header file? 

Honestly not sure what the best practices is for testing in Zephyr/C when it comes to things like this

jhedberg replied:
Seems like that concern should be solvable with proper documentation. These APIs either way require explicitly enabling `CONFIG_BT_TESTING` before they can be used. We could additionally output a build warning (if we don't already) that the configuration should not be used in production.

jhedberg replied:
@alwa-nordic this conversation is still unresolved

alwa-nordic replied:
I welcome the "public" test header. Should I do add it?

Should its doxygen say `@file @internal`? That will prevent anything from that file from rendering in the public documentation, I think.

Thalley replied:
I'm not a fan of public testing headers. If this is for internal testing, then I think such headers are better placed in a non-public place. Placing a `testing.h` makes more sense if it's a testing API that is designed to be used for users (maybe that is the intention?). 

We actually recently moved the Mesh `testing.h` header file from the public include directory to internal.

jhedberg replied:
The main issue with putting them e.g. somewhere under `subsys/bluetooth/` is the then you need to play manual tricks with include lookup paths etc. An alternative idea that comes to mind is to have some `CONFIG_BT_PRIVATE_INCLUDES` option which adds `subsys/bluetooth/` to the lookup path.

Thalley replied:
Semantically it's odd to have an `internal` directory in the `public` directory :D 

FYI I'm not blocking these changes as I don't have a solution/alternative, but just wanted to voice my concerns about putting functions that aren't meant to be used by applications in the same location as our APIs. We don't want it to be perceived as a test framework/interface of any sort, right?

jhedberg commented on include/zephyr/bluetooth/testing.h:
We have include guards in every public header

Thalley commented on include/zephyr/bluetooth/testing.h:
```diff
@@ -0,0 +1,23 @@
+ *  function, and it will automatically become the event
+ *  listener.
+ *
+ *  @warning This may be not invoked if
+ *  @kconfig{CONFIG_BT_TESTING} is not set.
```
```suggestion
 *  @kconfig_dep{CONFIG_BT_TESTING}.
```

Thalley reviewed (approved):
As mentioned, I will not block this PR based on the testing API, but I do have my reservations about it.
